lib/Transforms/LowerPolynomialEval/Patterns.cpp

```diff
@@ -0,0 +1,201 @@
+#include "lib/Transforms/LowerPolynomialEval/Patterns.h"
+
+#include "lib/Dialect/Polynomial/IR/PolynomialOps.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"        // from @llvm-project
+#include "mlir/include/mlir/Transforms/WalkPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "lower-polynomial-eval"
+
+namespace mlir {
+namespace heir {
+
+using polynomial::EvalOp;
+using polynomial::TypedFloatPolynomialAttr;
+
+TypedAttr getScalarOrDenseAttr(Type tensorOrScalarType, APFloat value) {
+  return TypeSwitch<Type, TypedAttr>(tensorOrScalarType)
+      .Case<FloatType>(
+          [&](FloatType type) { return FloatAttr::get(type, value); })
+      .Case<ShapedType>(
+          [&](ShapedType type) { return DenseElementsAttr::get(type, value); })
+      .Default([](Type) { return nullptr; });
+}
+
+LogicalResult LowerViaHorner::matchAndRewrite(EvalOp op,
+                                              PatternRewriter& rewriter) const {
+  Type evaluatedType = op.getValue().getType();
+  ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+  b.setInsertionPoint(op);
+
+  LLVM_DEBUG(llvm::dbgs() << "evaluatedType: " << evaluatedType << "\n");
+
+  auto attr =
+      dyn_cast<polynomial::TypedFloatPolynomialAttr>(op.getPolynomialAttr());
+  if (!attr) return failure();
+
+  auto terms = attr.getValue().getPolynomial().getTerms();
+  int64_t maxDegree = terms.back().getExponent().getSExtValue();
+  const int degreeThreshold = 5;
+  if (!shouldForce() && maxDegree >= degreeThreshold) return failure();
+
+  auto monomialMap = attr.getValue().getPolynomial().getCoeffMap();
+  DenseMap<int64_t, TypedAttr> attributeMap;
+  for (auto& [key, monomial] : monomialMap) {
+    attributeMap[key] =
+        getScalarOrDenseAttr(evaluatedType, monomial.getCoefficient());
+  }
+
+  // Start with the coefficient of the highest degree term
+  Value result =
+      b.create<arith::ConstantOp>(evaluatedType, attributeMap[maxDegree]);
+
+  // Apply Horner's method, accounting for possible missing terms
+  auto x = op.getOperand();
+  for (int64_t i = maxDegree - 1; i >= 0; i--) {
+    // Multiply by x
+    result = b.create<arith::MulFOp>(result, x);
```

Comment:
quick comment: support we have a polynomial to evaluate that only has even integer powers. this code will generate the entire monomial basis and we'll be left with a bunch of odd x power terms, even though we could just fold those and simplify them to muls by x^2 powers

obviously this pass shouldn't handle that, but i'm curious if CSE / canonicalization will do that automagically?

Reply:
> if CSE / canonicalization will do that automagically

I don't think so. And I feel it would be better to more generally identify that repeated muls can be replaced with iterated squaring. Or else that this pass should just emit iterated squaring directly.